### DISCUSSION TITLE
request help: body_filter how to send request in apisix

### DISCUSSION THREAD

**Discussion by zhangweidongyyl** [apache/apisix#5677] [Show and tell]
### Issue description

i want send request in body_filter,based on last request returned Status to send different req

### Environment

- apisix version (cmd: `apisix version`):
- OS (cmd: `uname -a`):
- OpenResty / Nginx version (cmd: `nginx -V` or `openresty -V`):
- etcd version, if have (cmd: run `curl http://127.0.0.1:9090/v1/server_info` to get the info from server-info API):
- apisix-dashboard version, if have:
- the plugin runner version, if the issue is about a plugin runner (cmd: depended on the kind of runner):
- luarocks version, if the issue is about installation (cmd: `luarocks --version`):

**Comment by tzssangglass** [comment]
we can use a 0 delay  `ngx.timer.at`, see: https://github.com/openresty/lua-nginx-module#ngxtimerat

**Comment by zhangweidongyyl** [comment]
> we can use a 0 delay `ngx.timer.at`, see: https://github.com/openresty/lua-nginx-module#ngxtimerat

This doesn't seem to solve my problem
i use 

> ngx.header[conf.headers_arr[i]] = val 

 in header_filter.
i use 

> ngx.redirect(host) 

in body_filter
I don't know if it can solve this problem, but it can't bring headers 。

**Comment by tzssangglass** [comment]
I cannot got what you want to do.

BTW, ngx.redirect is unavailable in `body_filter_by_lua`, see: https://github.com/openresty/lua-nginx-module#ngxredirect

**Comment by zhangweidongyyl** [comment]
> I cannot got what you want to do.
> 
> BTW, ngx.redirect is unavailable in `body_filter_by_lua`, see: https://github.com/openresty/lua-nginx-module#ngxredirect

ngx.req.set_uri  it seem can  solve my problem

**Comment by tzssangglass** [comment]
As I've already said, if you want to send the request in the body_filter phase, you can use ngx.timer to do so,see: https://moonbingbing.gitbooks.io/openresty-best-practices/content/ngx_lua/timer.html

**Reply by DREZG** [reply]
I am already in 'ngx timer. at' of callback function sends the request and gets the response result. How can I get the result of the response outside the callback function? I need take the resonse result to ngx.arg[1].

**Reply by tzssangglass** [reply]
ngx.timer is asynchronous.

**Comment by tzssangglass** [comment]
> based on last request returned Status to send different req

you can last request returned Status get by `get_last_failure`, see:https://github.com/openresty/lua-resty-core/blob/master/lib/ngx/balancer.md#get_last_failure

**Comment by zhangweidongyyl** [comment]
> As I've already said, if you want to send the request in the body_filter phase, you can use ngx.timer to do so,see: https://moonbingbing.gitbooks.io/openresty-best-practices/content/ngx_lua/timer.html

yes it can slove my problem ,i write like this 

`function _M.body_filter(conf, ctx) 
          local function push_data(premature, uri, args, status)
             -- push the data uri, args, and status to the remote
             -- via ngx.socket.tcp or ngx.socket.udp
             -- (one may want to buffer the data in Lua a bit to
             -- save I/O operations)
         end
         local ok, err = ngx.timer.at(0, push_data,
                                      ngx.var.uri, ngx.var.args, ngx.header.status)
         if not ok then
             ngx.log(ngx.ERR, "failed to create timer: ", err)
             return
         end

end`

but my ngx.socket.tcp() always return 400 Bad Request,how can i reslove this problem

**Comment by zhangweidongyyl** [comment]
i use tcp.socket.connect always return 

use https://github.com/openresty/lua-nginx-module#tcpsockconnect example
failed to connect to google: www.google.com could not be resolved (10060: Operation timed out)

**Comment by tzssangglass** [comment]
> but my ngx.socket.tcp() always return 400 Bad Request,how can i reslove this problem

This is a matter of the request method or request parameters. It has to do with your own service.

> failed to connect to google: [www.google.com](http://www.google.com) could not be resolved (10060: Operation timed out)

This is where your local network environment is unable to resolve the domain name and you can use another domain name that can resolve.

**Comment by tzssangglass** [comment]
I will turn this issue into a discussion, which has nothing to do with the project.

**Comment by showcup** [comment]
你可以利用resty.http库来发发送同步的http请求，像下面这样：
```lua

local http = require "resty.http" -- 加载 lua-resty-http 库

httpc, err = http:new()                                         -- 创建 http 连接对象
    httpc:set_timeout(time)                                         -- 注意名字里有一个下划线
    httpc:set_timeout(connect_timeout,                              -- 同上
            send_timeout, read_timeout)
    local res, err = httpc:request_uri(                             -- 发送 HTTP 请求，默认是 GET
            'http://www.baidu.com',                                     -- 指定 IP 地址
            {path = '/echo',                                        -- 指定具体路径
             query = {name = 'chrono'}}                              -- 请求的参数，使用 Lua 表
    )
    if not res then                                                 -- 检查请求是否成功
        core.log.error("failed to request : ", err)
        return
    end
    core.log.info("args:", type(args))
    core.log.info("res.body：",res.body)                                               -- 输出响应体
end
local resp = ""
local ok, err = ngx.timer.at(0, push_data, ngx.var.args)
if not ok then
    ngx.log(ngx.ERR, "failed to create timer: ", err)
    return
end
ngx.arg[1] = "success"
ngx.arg[2] = true`

```
